options/rtdl/generic/linker.cpp

```diff
@@ -1131,12 +1137,31 @@ void Loader::submitObject(SharedObject *object) {
 void Loader::linkObjects() {
 	_buildTlsMaps();
 
-	// Promote objects to the global scope.
+	// Promote objects to the desired scope.
 	for(auto it = _linkBfs.begin(); it != _linkBfs.end(); ++it) {
-		if((*it)->globalRts)
+		// The object was already loaded into the desired scope.
+		if ((*it)->loadScope == _loadScope)
 			continue;
-		(*it)->globalRts = _linkRts;
-		_globalScope->appendObject(*it);
+
+		if ((*it)->globalRts == 0 && _loadScope == globalScope.get())
+			(*it)->globalRts = _linkRts;
+
+		if ((*it)->loadScope) {
+			// The object was already loaded into a different scope.
+			// This may happen if e.g we load DSO foo with RTLD_LOCAL, and then
+			// load DSO bar with RTLD_GLOBAL where bar depends on foo.
+			// In this case, foo and all its dependencies must be added to
+			// the global scope. Note that the localScope of foo itself
+			// does not change here (it is fixed at first load, see below).
+
+			_loadScope->appendObject(*it);
+
+			// TODO: Add dependencies to the _loadScope too, in the correct order.
```

Comment:
Failing silently here could lead to some nasty bugs. Could we maybe warn about it?

Reply:
is this that hard to implement even?

Reply:
It's a bit annoying as we have to do a breadth-first traversal of the dependencies, but let me give it a go.

Reply:
I've implemented this now and added a test for it.